### ISSUE TITLE
Improving Plugins.

### ISSUE THREAD

**Issue by sunshinejr** [Moya/Moya#1123]
Hey all! I'm here to present my idea of improving Plugins and continue (kinda deadly) discussion from #985. 

Right now plugins:
- Are limited to one base type that has many functions that are not used and most of the time empty (but still executed).
- Are hard to expand (as a result of point above).
- Have limited functionality (as a result of points above).
- Do many things that `endpointClosure` & `requestClosure` do as well - making it ambiguous for users to use & for us to recommend a good practice.

Below is my proposition on to how make all of the above go away. Main idea is to remove `closures`, `manager` etc. from `MoyaProvider` and replace it with plugins. The first version may just introduce everything we need to a migration and then expand it as we go. Maybe we can even extract something from `TargetType`? 🤔 

### Types

```swift
// Base type

public protocol Plugin {}
```

```swift
// Current plugin methods extracted into Plugin types

public protocol RequestDelegatePluginType: Plugin {
    func willSend(_ request: RequestType, target: TargetType)
    func didReceive(_ result: Result<Moya.Response, MoyaError>, target: TargetType)
}

public protocol ResponseProcessingPluginType: Plugin {
    func process(_ result: Result<Moya.Response, MoyaError>, target: TargetType) -> Result<Moya.Response, MoyaError>
}
```

```swift
// `endpointClosure` & `requestClosure` & `stubbingClosure` into Plugin types

public protocol EndpointPreparingPluginType: Plugin {
    func prepare<T: TargetType>(endpoint: Endpoint<T>, target: T) -> Endpoint<T>
}

public protocol RequestPreparingPluginType: Plugin {
    func prepare(_ request: URLRequest, target: TargetType) -> URLRequest
}

public protocol ResponseStubbingPluginType: Plugin { // we may also remove this one and try to change `ResponseProcessingPluginType` a little bit to allow stubbing behavior :)
    func stubBehavior<T: TargetType>(_ target: T) -> StubBehavior
}
``` 

```swift
// Authentication moved from Request to Plugin type too

public protocol AuthenticationPluginType: Plugin {
    func authenticate(usingCredential credential: URLCredential) -> Self
}
```

### Migration from current closures
We could persist current closures as a default plugins, e.g. this `DefaultEndpointClosure` translated:
```swift
class DefaultEndpointPreparingPlugin: EndpointPreparingPluginType {
    
    func prepare<T>(endpoint: Endpoint<T>, target: T) -> Endpoint<T> where T: TargetType {
        return Endpoint(
            url: URL(for: target).absoluteString,
            sampleResponseClosure: { .networkResponse(200, target.sampleData) },
            method: target.method,
            parameters: target.parameters,
            parameterEncoding: target.parameterEncoding
        )
    }
}
```

### Problem - How to store plugins?
We might want only one plugin for some types (e.g. `ResponseStubbingPluginType`) and unlimited for other types (e.g. `RequestDelegatePluginType`). Having multiple of stubbing plugins or request/endpoint preparing plugins creates undefined behavior. I thought that we may use a type for storing it (we would then have some sort of value dependent types):
```swift
final class PluginsConfiguration {
    
    let endpointPreparation: EndpointPreparingPluginType?
    let requestPreparation: RequestPreparingPluginType?
    let stubBehavior: ResponseStubbingPluginType?
    ...
    
    init(stubBehavior: ResponseStubbingPluginType? = nil, endpointPreparation: EndpointPreparingPluginType = DefaultEndpointPreparingPlugin()...) {}
}
```

But even tho it gives us comfort with multiplicity of plugins for now (and who knows what the future holds :D), this is not scalable. Let's say I would have one plugin having implementation of three protocols - I would have to type it 3 times. Now let's say we want to expand with plugins - we may end up with 20 properties and counting.

The second solution would be just having an array of `Plugin` type in `MoyaProvider.init`. Now we can choose to e.g. make `asserts` in code to only have X number of plugins with given type. This is not the best, but this way we are sure we do not end up with undefined behavior.

If you guys have any other ideas it would be great to hear them. Also, if you are still reading this, thank you! This might be a big change, so I'd love every feedback I can get. If the whole idea is just not the right one for your taste - just say it. 

/cc @Moya/contributors 

Cheers! 🎉

**Comment by pedrovereza** [comment]
Cool topic, thanks for bringing it up @sunshinejr!

I have a few things I'd like to bring up:

* I'm a bit concerned that the base protocol of plugins is empty. I'd rather have a protocol that explictly exposes what it can be used for, or no base protocol at all.
* Maybe as a symptom of the above, all plugins you used as example define different contracts. for instance, what do an `EndpointPreparingPluginType` and `ResponseStubbingPluginType` have in common? What's the benefit of having both as `Plugin`?
* Not sure about the second storing options. Having an array of `Plugin` would probably force us to know the exact type of plugin before using it, making the `Plugin` abstraction not so useful.

I think we first need to have a clear vision of what exaclty is a plugin for Moya. Is is something external to the core API, similar to our current `PluginType`, which deals only with request/response? Or is it something capable of changing the internal behavior of the framework (like `EndpointPreparingPluginType` and `RequestPreparingPluginType`)? Or both?

**Comment by sunshinejr** [comment]
Thank you so much for the feedback, @pedrovereza 👌 

About empty base protocol - it would be needed if we wanted to go with solution 2 to the problem - array of plugins. For the first solution it would be possible to avoid the base protocol if there wouldn't be a reason for keeping it.

About different contracts - I agree that these are different. But [we already have different contracts in our current PluginType](https://github.com/Moya/Moya/blob/master/Sources/Moya/Plugin.swift#L10-L22) and I've just built on top of that. Maybe we need to think if we still want multiple contracts, but I believe that it would open more possibilities, like e.g. eliminating closures.

About second option to storing problem - we would need to have collections/properties of given plugins anyways, because we need to know which plugin to invocate, e.g. before firing up the request, we could invocate some plugin type that can decide if the request should use network or use stubs.

And the last but not least - what should our Plugin be? Given the fact how we use them right now (both delegation and modification of some objects), I wouldn't shrink the capabilities, rather stick to them (or stretch if needed). This way we would be able to reduce some responsibilities of Moya core and extract them into plugins. We might need to be circumspect in selecting which plugins we can add and which we cannot, though.

**Comment by colinta** [comment]
I have never used or written a Moya plugin, I'd like to hear what people are using them for (logging/debugging I suppose?).  But from a purely swift language point of view, wouldn't it be enough to have default implementations of the `Plugin` methods via an extension?  Having all these plugins is useful when you only want to write one method/feature, but more verbose and tedious if you wanted to implement more than that.

If we provided default/empty implementations of the `Plugin` methods then plugin authors could just write the methods they want/need, and the API would be simpler.

**Comment by colinta** [comment]
Shoot, I just looked up `Plugin.swift` and I see that it *already* implements empty/default methods.  So we've got that already.

I'm re-reading everything now.

**Comment by colinta** [comment]
After looking at `Plugins.swift`, I agree that we could revisit what we are doing with plugins.  But I'm not yet convinced that what you're proposing here is a step forward, it feels like a side step.  The reason I think that is because most of these methods are already in PluginType; all except the request and endpoint methods.

Using a "plugin" to implement those closures introduces a point of confusion for me - is it just *one* plugin that should create those?  What if multiple plugins can create the endpoint?  The example you provided has an `endpoint` argument, where does that come from?  What would it be used for in that plugin?  I don't think this is a compelling use case to do away with the current closure-based API.

I think we should attack this from a different point of view - instead of refactoring (into what is basically an equivalent system, but not backwards compatible), let's take a look at what people are trying to accomplish with plugins, and make sure we are allowing those use cases.

In short: I love the idea of refactoring plugins to make sure they are serving their purpose!  Do we have any more examples of that?

**Comment by sunshinejr** [comment]
@colinta you could use one plugin that implements few protocols, or just use multiple plugins - your choice. The point of this idea was to refactor plugins to have the same functionality as now, but also be easily expandable to features like `endpointClosure` & `requestClosure` or authentication/stubbing/inflights. 

Protocols in my example are just stubs for what we can really do - the `endpoint` function could get the parameter from the other plugin that makes an `Endpoint` from `TargetType`.

Nonetheless, If anyone have a different idea I'd love to hear it.

**Comment by sunshinejr** [comment]
Because of scale of this improvement and lack of feedback/better alternatives, I'm closing this one. Maybe we will revisit it some time in the future.
